Update dotnet core install and configure docs (#9641)
DataDog/documentation: content/en/tracing/setup_overview/setup/dotnet-core.md

      | sudo tar xzf - -C /opt/datadog
      ```
    - **Other distributions** - Download the tar archive with the glibc-linked binary:
      ```bash
      sudo mkdir -p /opt/datadog
      curl -L https://github.com/DataDog/dd-trace-dotnet/releases/download/v<TRACER_VERSION>/datadog-dotnet-apm-<TRACER_VERSION>.tar.gz \
      | sudo tar xzf - -C /opt/datadog
      ```
3. Add the [required environment variables](#required-environment-variables).

4. Run the `/opt/datadog/createLogPath.sh` script, which creates a directory for the log files and sets appropriate directory permissions. The default directory for log files is `/var/log/datadog/dotnet`.

5. Create application load.

6. Visit [APM Live Traces][3].

[1]: /agent/basic_agent_usage/
[2]: https://en.wikipedia.org/wiki/Musl
[3]: https://app.datadoghq.com/apm/traces
{{% /tab %}}

{{< /tabs >}}

By default, tracing is enabled in your `datadog.yaml` file under `apm_enabled: true` and the Agent listens for trace traffic at `localhost:8126`.

### Required environment variables

{{< tabs >}}

{{% tab "Windows" %}}

#### Applications not hosted in IIS

> To enable automatic instrumentation on For Windows applications not in IIS, you must set two environment variables before starting your application:

Reviewer: ```suggestion
To enable automatic instrumentation for Windows applications not hosted in IIS, you must set two environment variables before starting your application:
```

Author: heh, you can really tell I've been looking at these too long and no longer reading them. Thank you.

Name                       | Value
---------------------------|------
`CORECLR_ENABLE_PROFILING` | `1`
`CORECLR_PROFILER`         | `{846F5F1C-F9AE-4B07-969E-05C26BC060D8}`

<div class="alert alert-warning"> 
  <strong>Note:</strong> The .NET runtime tries to load a profiler into any .NET process started with these environment variables set. You should limit instrumentation only to the applications that need to be traced. Don't set these environment variables globally because this causes all .NET processes on the host to load the profiler.
</div>

> #### Windows services

Reviewer: "Windows services" falls under "Applications not hosted in IIS" above. Should it use `#####`?
```suggestion
##### Windows services
```

To automatically instrument a Windows service, set the `CORECLR_ENABLE_PROFILING` and `CORECLR_PROFILER` environment variables: 

1. In the Windows Registry Editor, create a multi-string value called `Environment` in the `HKLM\System\CurrentControlSet\Services\<SERVICE NAME>` key. 
2. Set the value data to:
    ```text
    CORECLR_ENABLE_PROFILING=1
    CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
    ```
    {{< img src="tracing/setup/dotnet/RegistryEditorCore.png" alt="Creating the environment variables for instrumenting a service in the Registry Editor" >}}

Alternatively, you can set the environment variables by using the following PowerShell snippet:

```powershell
[String[]] $v = @("CORECLR_ENABLE_PROFILING=1", "CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}")
Set-ItemProperty HKLM:SYSTEM\CurrentControlSet\Services\<NAME> -Name Environment -Value $v
```

> #### Console applications

Reviewer: "Console applications" falls under "Applications not hosted in IIS" above. Should it use `#####`?
```suggestion
##### Console applications
```

    text: "Examples of Custom Instrumentation"
---

## Compatibility requirements

The .NET Tracer supports instrumentation on:
  - .NET 5
  - .NET Core 3.1
  - .NET Core 2.1. 

For a full list of supported libraries, visit the [Compatibility Requirements][1] page.

## Automatic instrumentation

<div class="alert alert-warning"> 
  <strong>Note:</strong>  If you are using both automatic and custom instrumentation, it is important to keep the package versions (for example, MSI and NuGet) in sync.
</div>

### Installation

> By default, tracing is enabled in your `datadog.yaml` file under `apm_enabled: true` and the Agent listens for trace traffic at `localhost:8126`.

Reviewer: I think having this as the first line is confusing. It refers to Agent settings, but this section is about Tracer installation. There is not `datadog.yaml` in the Tracer. Did we intentionally remove the section "Configure the Datadog Agent for APM" below, @jaycdave88?

Reviewer 2: @lucaspimentel we did in order to focus on a more steps rather then large texts. Kari and I chatted on this - I was for removing since it doesnt exist today in our public docs for .NET Framework.

Reviewer: https://docs.datadoghq.com/tracing/setup_overview/setup/dotnet-framework#configure-the-datadog-agent-for-apm

Reviewer 2: Ah, thank you for the flag. 

Yes - lets just add that same section to our .NET Core docs. @kayayarai

{{< partial name="apm/apm-containers.html" >}}
</br>

> 3. After having instrumented your application, the tracing client sends traces to `localhost:8126` by default.  If this is not the correct host and port, change it by setting these environment variables:

Reviewer: ```suggestion
3. After instrumenting your application, the tracing client sends traces to `localhost:8126` by default.  If this is not the correct host and port, change it by setting these environment variables:
```
or
```
While it is instrumenting your application...
```

    - `DD_AGENT_HOST`

>     - `DD_TRACE_AGENT_PORT`

Reviewer: I don't know how these got here, but `DD_AGENT_HOST` and `DD_TRACE_AGENT_PORT` are optional settings that apply to all instrumentation and they are explained below in the configuration section. This section is only for environment variables required to enable automatic instrumentation.

Reviewer: Related: the first point above regarding `apm_non_local_traffic: true` only applies then the Agent is running in a different host or container. It's better suited for the "Linux Docker container" section below.

Author: I can move the `apm_non_local_traffic: true` step to the Linux Docker container for sure.

The two settings were in the docs in this step previously -- they have not been added in this edit -- and I think the "if this is not the correct host and port" correctly describes when should be used and when not, right?

Reviewer 2: @lucaspimentel - taking a look at how other languages handle the wording, this is how Java handles it today as well. Let's move forward with these items to keep the tracer docs aligned - if we think we can do better we can create smaller PR to fix.

Reviewer: @jaycdave88: Java has `apm_non_local_traffic` and `DD_AGENT_HOST`/`DD_TRACE_AGENT_PORT`, under the [Containers](https://docs-staging.datadoghq.com/kari/update-dotnet-core/tracing/setup_overview/setup/java?tab=containers#configure-the-datadog-agent-for-apm) section.

Other tracer do not have a section for "Required environment variables." This is specific to .NET because `CORECLR_ENABLE_PROFILING`/`CORECLR_PROFILER` is the only way to attach a profiler.

Reviewer: But yeah, moving `DD_AGENT_HOST` and `DD_TRACE_AGENT_PORT` in a separate PR is fine since it wasn't added in this PR.

dotnet example.dll
```

#### Linux Docker container

> To set the required environment variables on a Linux Docker container:

Reviewer 3: It feels like it would be nice to have an encapsulated Docker image example that shows how to do everything. i.e. 

* Install Tracer
* Run `/opt/datadog/createLogPath.sh`
* Add environment variables

Currently you have to try and piece it all together for your environment, instead of just having e.g. 3 example Docker files (for debian/alpine/ubuntu - [the Docker Image base-OSs supported by Microsoft](https://hub.docker.com/_/microsoft-dotnet-aspnet/))

Maybe it makes more sense to have those as blog/post sample? But that's what I would be looking for when trying to get up and running quickly

Reviewer 2: That is a good call out - maybe a good happy path would be to add an example Dockerfile to our Github / samples repo - then cross link this back? I am not sure about this one. @lucaspimentel thoughts?

Reviewer: We have something... somewhere. Can't remember right now, but it's definitely not linked in the docs, which is a good idea.

Author: I'm going to go ahead and publish this PR without this, and when you're ready to add an example, we can open a clean PR.

    systemctl set-environment DD_INTEGRATIONS=/opt/datadog/integrations.json
    systemctl set-environment DD_DOTNET_TRACER_HOME=/opt/datadog
    ```
2. Verify that the environment variables were set by running `systemctl show-environment`.

Once this is set, verify that the environment variables were set with `systemctl show-environment`.
3. Restart the .NET service for the environment variables to take effect.

[1]: /agent/guide/agent-configuration-files/#agent-main-configuration-file
[2]: https://docs.docker.com/engine/reference/builder/#env
[3]: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=
[2]: https://www.freedesktop.org/software/systemd/man/systemctl.html#set-environment%20VARIABLE=VALUE%E2%80%A6
{{% /tab %}}

{{< /tabs >}}

**Note:** The .NET runtime tries to load a profiler into _any_ .NET process that is started with these environment variables set. You should limit instrumentation only to the applications that need to be traced. **We do not recommend setting these environment variables globally as this causes _all_ .NET processes on the host to load the profiler.**

### Configure the Datadog Agent for APM

> Install and configure the Datadog Agent to receive traces from your instrumented application. By default the Datadog Agent is enabled in your `datadog.yaml` file under `apm_enabled: true` and listens for trace traffic at `localhost:8126`. For containerized environments, follow the in-app [Quickstart instructions][3] to enable trace collection within the Datadog Agent.

Reviewer: This links to https://docs.datadoghq.com/agent/basic_agent_usage/windows/, which is not "in-app Quickstart instructions." We either have the wrong link or the wrong description.

#### SystemCTL (all services)

<div class="alert alert-warning"> 
  <strong>Note:</strong> The .NET runtime tries to load a profiler into <em>any</em> .NET process that is started with these environment variables set. You should limit instrumentation to only the applications that need to be traced. <strong>Don't set these environment variables globally as this causes <em>all</em> .NET processes on the host to load the profiler.</strong>
</div>

When using `systemctl` to run .NET applications as a service, you can also set environment variables to be loaded for all services run by `systemctl`. 

1. Set the required environment variables by running [`systemctl set-environment`][2]:

    ```bat
    systemctl set-environment CORECLR_ENABLE_PROFILING=1
    systemctl set-environment CORECLR_PROFILER={846F5F1C-F9AE-4B07-969E-05C26BC060D8}
    systemctl set-environment CORECLR_PROFILER_PATH=/opt/datadog/Datadog.Trace.ClrProfiler.Native.so
    systemctl set-environment DD_INTEGRATIONS=/opt/datadog/integrations.json
    systemctl set-environment DD_DOTNET_TRACER_HOME=/opt/datadog
    ```
2. Verify that the environment variables were set by running `systemctl show-environment`.

> 3. Restart the .NET service for the environment variable settings to take effect.

Reviewer: ```suggestion
3. Restart the .NET service for the environment variables to take effect.
```

Tracer.Instance = tracer;
```

{{% /tab %}}

{{% tab "JSON file" %}}

> To configure the Tracer using a JSON file, create `datadog.json` in the instrumented application's directory. The root JSON object must be hashed with a key/value pair for each setting. For example:

Reviewer: "A hash" is used colloquially in JSON to refer to objects (because they behave like hash tables). Let's change "a hash" to "an object."

```suggestion
To configure the Tracer using a JSON file, create `datadog.json` in the instrumented application's directory. The root JSON object must be an object with a key/value pair for each setting. For example:
```

|----------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `DD_TRACE_ENABLED`<br/><br/>`TraceEnabled`                      | Enables or disables all automatic instrumentation. Setting the environment variable to `false` completely disables the CLR profiler. For other configuration methods, the CLR profiler is still loaded, but traces will not be generated. Valid values are: `true` (default) or `false`. |
| `DD_TRACE_LOG_DIRECTORY`                                        | Sets the directory for .NET Tracer logs.<br/><br/>Default: `%ProgramData%\Datadog .NET Tracer\logs\`           |

> | `DD_TRACE_LOG_PATH`                                             | Sets the path for the automatic instrumentation log file and determines the directory of all other .NET Tracer log files. Ignored if `DD_TRACE_LOG_DIRECTORY` is set.             |

Reviewer: Custom instrumentation can also generates log. These `DD_TRACE_LOG_*` settings should be moved to the table above.
